### PR TITLE
RGW/logging: add rest API to flush logging bucket

### PR DESCRIPTION
**Pull request by AliMasarweh**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/rgw/rgw_rest_bucket_logging.cc [hunk 1]
```diff
@@ -209,6 +209,122 @@ class RGWPutBucketLoggingOp : public RGWDefaultResponseOp {
...
+  RGWOpType get_type() override { return RGW_OP_POST_BUCKET_LOGGING; }
+  uint32_t op_mask() override { return RGW_OP_TYPE_WRITE; }
+
+  void execute(optional_yield y) override {
+    op_ret = verify_bucket_logging_params(this, s);
+    if (op_ret < 0) {
+      return;
+    }
+
+    const auto max_size = s->cct->_conf->rgw_max_put_param_size;
```
Comment (yuvalif) [inline src/rgw/rgw_rest_bucket_logging.cc:237]: i don't think we expect any XML payload in this operation. wht information do we need except the source bucket name (which is in the request not the XML payload)?

File: src/rgw/rgw_rest_bucket_logging.cc [hunk 2]
```diff
@@ -209,6 +209,122 @@ class RGWPutBucketLoggingOp : public RGWDefaultResponseOp {
...
+      return;
+    }
+    rgw::bucketlogging::configuration configuration;
+    try {
+      configuration.enabled = true;
+      decode(configuration, iter->second);
+    } catch (buffer::error& err) {
+      ldpp_dout(this, 1) << "ERROR: failed to decode logging attribute '" << RGW_ATTR_BUCKET_LOGGING
+                         << "'. error: " << err.what() << dendl;
+      return;
```
Comment (yuvalif) [inline src/rgw/rgw_rest_bucket_logging.cc:296]: do we expect result here to be sucess?
if not we should probably set:`op_ret = -EINVAL`

**Comment by yuvalif** [comment]
LGTM. 
currently in the process of testing https://github.com/ceph/ceph/pull/59808
so, we should probably take this work after the main PR is merged

File: src/rgw/rgw_rest_s3.cc [hunk 3]
```diff
@@ -4921,6 +4912,10 @@ RGWOp *RGWHandler_REST_Bucket_S3::op_post()
     return new RGWDeleteMultiObj_ObjStore_S3;
   }
 
+  if (s->info.args.sub_resource_exists("logging")) {
```
Comment (yuvalif) [inline src/rgw/rgw_rest_s3.cc:4915]:
```suggestion
  if (s->info.args.exists("logging")) {
```

File: examples/rgw/boto3/post_bucket_logging.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+        """post bucket logging for flushing"""
+        parameters = {'logging': None}
+        response, status = self.send_request('POST', parameters)
+        dict_response = xmltodict.parse(response, parameters)
+        return dict_response, status
+
+
+def connection():
+    # endpoint and keys from vstart
+    hostname = 'http://127.0.0.1'
```
Comment (yuvalif) [inline examples/rgw/boto3/post_bucket_logging.py:82]: this is url not hostname

File: src/test/rgw/bucket_notification/api.py [hunk 5]
```diff
@@ -230,6 +230,32 @@ def del_config(self, notification=None):
 
 test_path = os.path.normpath(os.path.dirname(os.path.realpath(__file__))) + '/../'
 
+
+class BucketLogging:
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/api.py:234]: we don't need this.
bucket logging tests are done elsewhere (as part of s3test)

**Comment by yuvalif** [comment]
s3tests added here: https://github.com/ceph/s3-tests/pull/606

**Comment by yuvalif** [comment]
please add documentation both to bucket ops doc and bucket logging doc.

**Comment by AliMasarweh** [comment]
s3tests passed

**Comment by AliMasarweh** [comment]
> s3tests added here: [ceph/s3-tests#606](https://github.com/ceph/s3-tests/pull/606)

s3tests for bucket logging passed

**Review by yuvalif — CHANGES_REQUESTED** [review]
please remove the test branch commit before merge

**Comment by AliMasarweh** [comment]
> please remove the test branch commit before merge

Done

**Review by yuvalif — APPROVED** [review]

**Comment by yuvalif** [comment]
@AliMasarweh please merge once all tests are passing

**Comment by AliMasarweh** [comment]
jenkins test make check

**Pull request merged**